From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] wmp: IWMPPlayer4::playlistCollection and IWMPPlaylistCollection

Stub (E_NOTIMPL + FIXME), as documented now.

get_playlistCollection returns an embedded object. There is no playlist
store: getAll/getByName return an empty IWMPPlaylistArray (count 0, Item is
E_INVALIDARG), newPlaylist returns a named empty playlist that is not stored,
isDeleted is VARIANT_FALSE, setDeleted/remove are S_FALSE, importPlaylist hands
back the playlist passed in. NULL out pointers/items are E_POINTER, NULL names
E_INVALIDARG.
Gate: test/wmp-collections-gate.sh.

---
diff --git a/dlls/wmp/player.c b/dlls/wmp/player.c
--- a/dlls/wmp/player.c
+++ b/dlls/wmp/player.c
@@ -419,8 +419,13 @@ static HRESULT WINAPI WMPPlayer4_get_mediaCollection(IWMPPlayer4 *iface, IWMPMed
 static HRESULT WINAPI WMPPlayer4_get_playlistCollection(IWMPPlayer4 *iface, IWMPPlaylistCollection **ppPlaylistCollection)
 {
     WindowsMediaPlayer *This = impl_from_IWMPPlayer4(iface);
-    FIXME("(%p)->(%p)\n", This, ppPlaylistCollection);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, ppPlaylistCollection);
+
+    if (!ppPlaylistCollection)
+        return E_POINTER;
+    IWMPPlaylistCollection_AddRef(*ppPlaylistCollection = &This->IWMPPlaylistCollection_iface);
+    return S_OK;
 }
 
 static HRESULT WINAPI WMPPlayer4_get_versionInfo(IWMPPlayer4 *iface, BSTR *version)
@@ -3470,6 +3475,180 @@ static const IWMPMediaCollectionVtbl WMPMediaCollectionVtbl = {
     WMPMediaCollection_isDeleted,
 };
 
+typedef struct {
+    IWMPPlaylistArray IWMPPlaylistArray_iface;
+    LONG ref;
+} WMPPlaylistArray;
+
+static HRESULT WINAPI WMPPlaylistArray_QueryInterface(IWMPPlaylistArray *iface, REFIID riid, void **ppv)
+{
+    if (IsEqualGUID(riid, &IID_IUnknown) || IsEqualGUID(riid, &IID_IDispatch)
+            || IsEqualGUID(riid, &IID_IWMPPlaylistArray))
+    {
+        *ppv = iface;
+        IUnknown_AddRef((IUnknown *)iface);
+        return S_OK;
+    }
+    WARN("Unsupported interface (%s)\n", wine_dbgstr_guid(riid));
+    *ppv = NULL;
+    return E_NOINTERFACE;
+}
+
+static ULONG WINAPI WMPPlaylistArray_AddRef(IWMPPlaylistArray *iface)
+{
+    WMPPlaylistArray *This = CONTAINING_RECORD(iface, WMPPlaylistArray, IWMPPlaylistArray_iface);
+    return InterlockedIncrement(&This->ref);
+}
+
+static ULONG WINAPI WMPPlaylistArray_Release(IWMPPlaylistArray *iface)
+{
+    WMPPlaylistArray *This = CONTAINING_RECORD(iface, WMPPlaylistArray, IWMPPlaylistArray_iface);
+    ULONG ref = InterlockedDecrement(&This->ref);
+    if (!ref)
+        free(This);
+    return ref;
+}
+
+WMP_IDISPATCH_METHODS(WMPPlaylistArray, IWMPPlaylistArray, IWMPPlaylistArray_tid)
+
+static HRESULT WINAPI WMPPlaylistArray_get_count(IWMPPlaylistArray *iface, LONG *plCount)
+{
+    TRACE("(%p)->(%p)\n", iface, plCount);
+    return empty_count(plCount);
+}
+
+static HRESULT WINAPI WMPPlaylistArray_Item(IWMPPlaylistArray *iface, LONG lIndex, IWMPPlaylist **ppItem)
+{
+    TRACE("(%p)->(%ld %p)\n", iface, lIndex, ppItem);
+    if (!ppItem)
+        return E_POINTER;
+    *ppItem = NULL;
+    return E_INVALIDARG;
+}
+
+static const IWMPPlaylistArrayVtbl WMPPlaylistArrayVtbl = {
+    WMPPlaylistArray_QueryInterface,
+    WMPPlaylistArray_AddRef,
+    WMPPlaylistArray_Release,
+    WMPPlaylistArray_GetTypeInfoCount,
+    WMPPlaylistArray_GetTypeInfo,
+    WMPPlaylistArray_GetIDsOfNames,
+    WMPPlaylistArray_Invoke,
+    WMPPlaylistArray_get_count,
+    WMPPlaylistArray_Item,
+};
+
+static HRESULT create_playlist_array(IWMPPlaylistArray **out)
+{
+    WMPPlaylistArray *array;
+
+    *out = NULL;
+    if (!(array = calloc(1, sizeof(*array))))
+        return E_OUTOFMEMORY;
+    array->IWMPPlaylistArray_iface.lpVtbl = &WMPPlaylistArrayVtbl;
+    array->ref = 1;
+    *out = &array->IWMPPlaylistArray_iface;
+    return S_OK;
+}
+
+
+static HRESULT WINAPI WMPPlaylistCollection_newPlaylist(IWMPPlaylistCollection *iface, BSTR name,
+        IWMPPlaylist **ppItem)
+{
+    TRACE("(%p)->(%s %p)\n", iface, debugstr_w(name), ppItem);
+
+    if (!ppItem)
+        return E_POINTER;
+    *ppItem = NULL;
+    if (!name || !*name)
+        return E_INVALIDARG;
+    /* the playlist is not stored: getAll()/getByName() never list it */
+    return create_playlist(name, NULL, 0, ppItem);
+}
+
+static HRESULT WINAPI WMPPlaylistCollection_getAll(IWMPPlaylistCollection *iface, IWMPPlaylistArray **ppArray)
+{
+    TRACE("(%p)->(%p)\n", iface, ppArray);
+
+    if (!ppArray)
+        return E_POINTER;
+    return create_playlist_array(ppArray);
+}
+
+static HRESULT WINAPI WMPPlaylistCollection_getByName(IWMPPlaylistCollection *iface, BSTR name,
+        IWMPPlaylistArray **ppArray)
+{
+    TRACE("(%p)->(%s %p)\n", iface, debugstr_w(name), ppArray);
+
+    if (!ppArray)
+        return E_POINTER;
+    *ppArray = NULL;
+    if (!name)
+        return E_INVALIDARG;
+    return create_playlist_array(ppArray);
+}
+
+/* a playlist that is not in the (empty) collection cannot be removed */
+static HRESULT WINAPI WMPPlaylistCollection_remove(IWMPPlaylistCollection *iface, IWMPPlaylist *pItem)
+{
+    TRACE("(%p)->(%p)\n", iface, pItem);
+    return pItem ? S_FALSE : E_POINTER;
+}
+
+static HRESULT WINAPI WMPPlaylistCollection_setDeleted(IWMPPlaylistCollection *iface, IWMPPlaylist *pItem,
+        VARIANT_BOOL fIsDeleted)
+{
+    TRACE("(%p)->(%p %d)\n", iface, pItem, fIsDeleted);
+    return pItem ? S_FALSE : E_POINTER;
+}
+
+static HRESULT WINAPI WMPPlaylistCollection_isDeleted(IWMPPlaylistCollection *iface, IWMPPlaylist *pItem,
+        VARIANT_BOOL *pfIsDeleted)
+{
+    TRACE("(%p)->(%p %p)\n", iface, pItem, pfIsDeleted);
+
+    if (!pfIsDeleted)
+        return E_POINTER;
+    *pfIsDeleted = VARIANT_FALSE;
+    return pItem ? S_OK : E_POINTER;
+}
+
+/* importing hands back the playlist that was passed in */
+static HRESULT WINAPI WMPPlaylistCollection_importPlaylist(IWMPPlaylistCollection *iface, IWMPPlaylist *pItem,
+        IWMPPlaylist **ppImportedItem)
+{
+    TRACE("(%p)->(%p %p)\n", iface, pItem, ppImportedItem);
+
+    if (!ppImportedItem)
+        return E_POINTER;
+    *ppImportedItem = NULL;
+    if (!pItem)
+        return E_POINTER;
+    IWMPPlaylist_AddRef(*ppImportedItem = pItem);
+    return S_OK;
+}
+
+WMP_EMBEDDED_UNKNOWN(WMPPlaylistCollection, IWMPPlaylistCollection, IID_IWMPPlaylistCollection,
+        IWMPPlaylistCollection_iface)
+WMP_IDISPATCH_METHODS(WMPPlaylistCollection, IWMPPlaylistCollection, IWMPPlaylistCollection_tid)
+
+static const IWMPPlaylistCollectionVtbl WMPPlaylistCollectionVtbl = {
+    WMPPlaylistCollection_QueryInterface,
+    WMPPlaylistCollection_AddRef,
+    WMPPlaylistCollection_Release,
+    WMPPlaylistCollection_GetTypeInfoCount,
+    WMPPlaylistCollection_GetTypeInfo,
+    WMPPlaylistCollection_GetIDsOfNames,
+    WMPPlaylistCollection_Invoke,
+    WMPPlaylistCollection_newPlaylist,
+    WMPPlaylistCollection_getAll,
+    WMPPlaylistCollection_getByName,
+    WMPPlaylistCollection_remove,
+    WMPPlaylistCollection_setDeleted,
+    WMPPlaylistCollection_isDeleted,
+    WMPPlaylistCollection_importPlaylist,
+};
+
 static HRESULT WINAPI WMPError_QueryInterface(IWMPError *iface, REFIID riid, void **ppv)
 {
     if (IsEqualGUID(riid, &IID_IUnknown) || IsEqualGUID(riid, &IID_IDispatch) || IsEqualGUID(riid, &IID_IWMPError))
@@ -3731,6 +3910,7 @@ BOOL init_player(WindowsMediaPlayer *wmp)
     wmp->IWMPError_iface.lpVtbl = &WMPErrorVtbl;
     wmp->IWMPClosedCaption_iface.lpVtbl = &WMPClosedCaptionVtbl;
     wmp->IWMPMediaCollection_iface.lpVtbl = &WMPMediaCollectionVtbl;
+    wmp->IWMPPlaylistCollection_iface.lpVtbl = &WMPPlaylistCollectionVtbl;
 
     name = SysAllocString(L"Playlist1");
     if (SUCCEEDED(create_playlist(name, NULL, 0, &playlist)))
diff --git a/dlls/wmp/wmp_private.h b/dlls/wmp/wmp_private.h
--- a/dlls/wmp/wmp_private.h
+++ b/dlls/wmp/wmp_private.h
@@ -33,6 +33,8 @@
     XIID(IWMPPlaylist) \
     XIID(IWMPError) \
     XIID(IWMPClosedCaption) \
+    XIID(IWMPPlaylistCollection) \
+    XIID(IWMPPlaylistArray) \
     XIID(IWMPMediaCollection) \
     XIID(IWMPStringCollection) \
     CTID(WindowsMediaPlayer)
@@ -117,6 +119,7 @@ struct WindowsMediaPlayer {
     IWMPNetwork IWMPNetwork_iface;
     IWMPError IWMPError_iface;
     IWMPClosedCaption IWMPClosedCaption_iface;
+    IWMPPlaylistCollection IWMPPlaylistCollection_iface;
     IWMPMediaCollection IWMPMediaCollection_iface;
 
     LONG ref;
